mypyc/cstring.py

```diff
@@ -0,0 +1,53 @@
+"""Encode valid C string literals from Python strings.
+
+If a character is not allowed in C string literals, it is either emitted
+as a simple escape sequence (e.g. '\\n'), or an octal escape sequence
+with exactly three digits ('\\oXXX'). Question marks are escaped to
+prevent trigraphs in the string literal from being interpreted (this
+escape sequence is invalid in Python).
+
+Consider the string literal 'AB\\xCDEF'. As one would expect, Python
+parses it as ['A', 'B', 0xCD, 'E', 'F']. However, the C standard
+specifies that all hexadecimal digits immediately following '\\x' will
+be interpreted as part of the escape sequence. Therefore, it is
+unexpectedly parsed as ['A', 'B', 0xCDEF].
+
+Emitting ('AB\\xCD' 'EF') would avoid this behaviour. However, we opt
+for simplicity and use octal escape sequences instead. They do not
+suffer from the same issue as they are defined to parse at most three
+octal digits.
+"""
+
+import string
+from typing import Tuple
+
+CHAR_MAP = ['\\{:03o}'.format(i) for i in range(256)]
+
+# It is safe to use string.printable as it always uses the C locale.
+for c in string.printable:
+    CHAR_MAP[ord(c)] = c
+
+# These assignments must come last because we prioritize simple escape
+# sequences over any other representation.
+CHAR_MAP[ord('\'')] = r'\''
+CHAR_MAP[ord('\"')] = r'\"'
+CHAR_MAP[ord('?')] = r'\?'  # This escape sequence is invalid in Python.
+CHAR_MAP[ord('\\')] = r'\\'
+CHAR_MAP[ord('\a')] = r'\a'
+CHAR_MAP[ord('\b')] = r'\b'
+CHAR_MAP[ord('\f')] = r'\f'
+CHAR_MAP[ord('\n')] = r'\n'
+CHAR_MAP[ord('\r')] = r'\r'
+CHAR_MAP[ord('\t')] = r'\t'
+CHAR_MAP[ord('\v')] = r'\v'
+
+
+def encode_as_c_string(s: str) -> Tuple[str, int]:
+    """Produce a quoted C string literal and its size, for a UTF-8 string."""
+    return encode_bytes_as_c_string(s.encode('utf-8'))
+
+
+def encode_bytes_as_c_string(b: bytes) -> Tuple[str, int]:
```

Comment:
This is what an implementation with hexadecimal escape sequences, instead of octal, would look like:

```diff
diff --git i/mypyc/cstring.py w/mypyc/cstring.py
index 37f3ab38..0557a216 100644
--- i/mypyc/cstring.py
+++ w/mypyc/cstring.py
@@ -19,9 +19,11 @@ octal digits.
 """
 
 import string
+import itertools
 from typing import Tuple
 
-CHAR_MAP = ['\\{:03o}'.format(i) for i in range(256)]
+HEX_DIGITS = frozenset(string.hexdigits)
+CHAR_MAP = {}
 
 # It is safe to use string.printable as it always uses the C locale.
 for c in string.printable:
@@ -49,5 +51,13 @@ def encode_as_c_string(s: str) -> Tuple[str, int]:
 
 def encode_bytes_as_c_string(b: bytes) -> Tuple[str, int]:
     """Produce a quoted C string literal and its size, for a byte string."""
-    escaped = ''.join([CHAR_MAP[i] for i in b])
+    escaped = ''
+    for printable, group in itertools.groupby(b, key=CHAR_MAP.__contains__):
+        if printable:
+            s = ''.join([CHAR_MAP[i] for i in group])
+            if len(escaped) and s[0] in HEX_DIGITS:
+                escaped += '" "'
+            escaped += s
+        else:
+            escaped += ''.join(['\\x{:02X}'.format(i) for i in group])
     return '"{}"'.format(escaped), len(b)
```